Sources/Confidence/ConfidenceAppLifecycleProducer.swift

```diff
@@ -0,0 +1,113 @@
+        let currentBuild: String = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? ""
+        let context: ConfidenceStruct = [
+            "version": .init(string: currentVersion),
+            "build": .init(string: currentBuild)
+        ]
+
+        self.currentProducedContext.send(context)
+    }
+
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+
+    public func produceEvents() -> AnyPublisher<Event, Never> {
+        events.compactMap { event in event }.eraseToAnyPublisher()
+    }
+
+    public func produceContexts() -> AnyPublisher<ConfidenceStruct, Never> {
+        currentProducedContext
+            .filter { context in !context.isEmpty }
+            .eraseToAnyPublisher()
+    }
+
+    private func track(eventName: String) {
+        let previousBuild: String? = UserDefaults.standard.string(forKey: Self.buildNameKey)
+        let previousVersion: String? = UserDefaults.standard.string(forKey: Self.versionNameKey)
+
+        let currentVersion: String = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
+        let currentBuild: String = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? ""
+
+        let message: ConfidenceStruct = [
+            "version": .init(string: currentVersion),
+            "build": .init(string: currentBuild)
+        ]
+
+        if eventName == "app-launched" && !isLaunched {
+            isLaunched = true
+            if previousBuild != currentBuild || previousVersion != currentVersion {
+                events.send(Event(name: "app-updated", message: message))
+                return
+            } else {
+                events.send(Event(name: "app-installed", message: message))
+                return
+            }
+        }
+
+        events.send(Event(name: eventName, message: message))
+
+        UserDefaults.standard.setValue(currentVersion, forKey: Self.versionNameKey)
+        UserDefaults.standard.setValue(currentBuild, forKey: Self.buildNameKey)
+    }
+
+    private func updateContext(isForeground: Bool) {
+        withLock { [weak self] in
+            guard let self = self else {
+                return
+            }
+            var context = self.currentProducedContext.value
+            context.updateValue(.init(boolean: isForeground), forKey: "is_foreground")
+            self.currentProducedContext.send(context)
+        }
+    }
+
+    private func withLock(callback: @escaping () -> Void) {
+        queue.sync {
+            callback()
+        }
+    }
+
+    @objc func notificationResponse(notification: NSNotification) {
+        switch notification.name {
+        case UIApplication.didEnterBackgroundNotification:
+            updateContext(isForeground: false)
+        case UIApplication.willEnterForegroundNotification:
+            updateContext(isForeground: true)
+        case UIApplication.didFinishLaunchingNotification:
+            track(eventName: "app-launched")
```

Comment:
I am not sure this works if we start listening to this notification after the app has launched

Reply:
I am having issues getting this event emitted properly (with the is_foreground context)